src/request.rs

```diff
@@ -238,27 +267,176 @@ impl RequestBuilder {
+    ///
+    /// The headers will be merged in to any already set.
+    pub fn headers(&mut self, headers: ::header::Headers) -> &mut MultipartRequestBuilder {
+        self.request_builder.request_mut().headers.extend(headers.iter());
+        self
+    }
+
+    /// Enable HTTP basic authentication.
+    pub fn basic_auth<U, P>(&mut self, username: U, password: Option<P>) -> &mut MultipartRequestBuilder
+        where U: Into<String>,
+              P: Into<String>
+    {
+        self.request_builder.header(::header::Authorization(::header::Basic {
+                                                username: username.into(),
+                                                password: password.map(|p| p.into()),
+                                            }));
+        self
+    }
+
+    /// Set the request body.
+    pub fn body<T: Into<Body>>(&mut self, body: T) -> &mut MultipartRequestBuilder {
+        self.request_builder.request_mut().body = Some(body.into());
+        self
+    }
+
+    /// Add parameters to the multipart/form-data `Request`.
+    pub fn params(&mut self, params: Params) -> &mut MultipartRequestBuilder {
+        self.params = Some(params);
+        self
+    }
+
+    /// Add files to the multipart/form-data `Request`.
+    pub fn files(&mut self, files: Files) -> &mut MultipartRequestBuilder {
+        self.files = Some(files);
+        self
+    }
+
+    /// Build a `Request`, which can be inspected, modified and executed with
+    /// `Client::execute()`.
+    ///
+    /// # Panics
+    ///
+    /// This method consumes builder internal state. It panics on an attempt to
+    /// reuse already consumed builder.
+    pub fn build(&mut self) -> ::Result<Request> {
+        let mut body: Vec<u8> = Vec::new();
+        let boundary = MultipartRequestBuilder::choose_boundary();
+        let multipart_mime = ContentType(format!{"multipart/form-data; boundary={}", boundary}
+                                             .parse::<mime::Mime>()
+                                             .unwrap());
+        if let Some(p) = self.params.clone() {
+            for (name, value) in p {
+                write_bytes!(body, "\r\n--{}\r\n", boundary);
+                write_bytes!(body, "Content-Disposition: form-data; name=\"{}\"\n", name);
+                write_bytes!(body, "\r\n{}", value);
+            }
+        }
+
+        if let Some(f) = self.files.clone() {
+            for File { name, path, mime } in f {
+                write_bytes!(body, "\r\n--{}\r\n", boundary);
+                write_bytes!(body, "Content-Disposition: form-data; name=\"{}\"", name);
+                write_bytes!(body,
+                             "; filename=\"{}\"",
+                             path.file_name().unwrap().to_str().unwrap());
+                write_bytes!(body, "\r\nContent-type: {}\r\n\r\n", mime);
+                let mut content = try_!(fs::File::open(path));
+                content.read_to_end(&mut body).unwrap();
```

Comment:
```rust
struct MultipartBody {
    cursor: SomethingToRecordWhatParamOrFile,
    params: Params,
    files: Vec<File>,
}

impl Read for MultipartBody {
    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
        let mut n = 0;
        if self.params_bytes_remaining() {
            // take &mut [u8], use (&mut [u8]).write(other_bytes))
            n += copy_param_bytes(buf);
        }
        while !buf.is_empty() && !self.files.is_empty() {
            // basically, self.files[self.file_pos].read(buf)
            // and then when files is EOF, self.file_pos += 1
            n += self.copy_file(buf)?
        }
        Ok(n)
    }
}
```

The point here is that instead of pushing into a `Vec` that might grow to huge sizes, bytes are only pushed into the read slice. After a `read` from the body, that slice is pushed onto the socket, and only afterwards is another `read` done, so the amount of memory used stays low.